services/datastore/go-redis.go

```diff
@@ -0,0 +1,241 @@
+	if err := Store.Client.Del(ctx, key).Err(); err != nil {
+		return redisError{errors.New("failed to delete key"), key}
+	}
+	return nil
+}
+
+// Returns a string array representation of the values at the given key
+func (Store *DB) GetSet(ctx context.Context, key string) ([]string, error) {
+	res := Store.Client.SMembers(ctx, key)
+	if err := res.Err(); err != nil {
+		return nil, redisError{errors.New("failed to get set at key"), key}
+	}
+	return res.Val(), nil
+}
+
+// Adds given value to the set at the given key
+func (Store *DB) AddToSet(ctx context.Context, key string, obj string) error {
+	if err := Store.Client.SAdd(
+		ctx,
+		key,
+		obj,
+	).Err(); err != nil {
+		return redisError{errors.New("failed to add to set at key"), key}
+	}
+	return nil
+}
+
+// Deletes the given value from the set at the given
+func (Store *DB) DeleteFromSet(ctx context.Context, key string, obj string) error {
+	if err := Store.Client.SRem(ctx, key, obj).Err(); err != nil {
+		return redisError{errors.New("failed to delete from set at key"), key}
+	}
+	return nil
+}
+
+// TODO: @cj-chung add tests and comments for functions below
+func (Store *DB) Set(ctx context.Context, key string, val interface{}) error {
+	jsonVal, err := json.Marshal(val)
+	if err != nil {
+		return err
+	}
+	err = Store.Client.Set(ctx, key, jsonVal, 0).Err()
+	if err != nil {
+		return redisError{errors.New("failed to append to list, key in use for a different data structure"), fmt.Sprintf("Key: %s Value: %s Error: %s", key, jsonVal, err.Error())}
+	}
+	return nil
+}
+
+func (Store *DB) Get(ctx context.Context, key string) (string, error) {
+	val, err := Store.Client.Get(ctx, key).Result()
+	if err != nil {
+		return "", redisError{errors.New("database GET failed"), fmt.Sprintf("Key: %s Error: %s", key, err.Error())}
+	}
+	return val, nil
+}
+
+// Function for finding an entity from database
+func (Store *DB) FindEntity(ctx context.Context, id string, prefix string, entity interface{}) error {
+	if id == "" {
+		return redisError{errors.New("invalid ID"), id}
+	}
+	value, err := Store.Client.Get(ctx, prefix+id).Result()
+	if err != nil {
+		return redisError{errors.New("failed to get"), prefix + id}
+	}
+	err = json.Unmarshal([]byte(value), &entity)
+	return redisError{errors.New("unmarshal failed"), fmt.Sprintf("ID: %s, Error: %s", prefix+id, err.Error())}
+}
+
+func (Store *DB) GetAllValuesWithKeyPrefix(ctx context.Context, prefix string) (map[string]string, error) {
+	keys, err := getKeys(ctx, fmt.Sprintf("%s*", prefix), Store)
+	if err != nil {
+		return nil, err
+	}
+
+	values, err := getKeyAndValueMap(ctx, keys, prefix, Store)
+	if err != nil {
+		return nil, err
+	}
+	return values, nil
+}
+
+func getKeys(ctx context.Context, prefix string, store *DB) ([]string, error) {
+	var allkeys []string
+
+	iter := store.Client.Scan(ctx, 0, prefix, 0).Iterator()
+	for iter.Next(ctx) {
+		allkeys = append(allkeys, iter.Val())
+	}
+
```

Comment:
@phudtran how to make sure these functions work?

Reply:
I ported these over from the scenario managers database exactly as they were.